drivers/frequency/adf5902/adf5902.c

```diff
@@ -0,0 +1,937 @@
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <malloc.h>
+#include <math.h>
+#include "adf5902.h"
+#include "error.h"
+#include "delay.h"
+#include "util.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * @brief Writes 4 bytes of data to ADF5902.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data value to write.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int32_t adf5902_write(struct adf5902_dev *dev, uint8_t reg_addr, uint32_t data)
+{
+	uint8_t buff[ADF5902_BUFF_SIZE_BYTES];
+
+	data = data | reg_addr;
+
+	buff[0] = data >> 24;
+	buff[1] = data >> 16;
+	buff[2] = data >> 8;
+	buff[3] = data;
+
+	return spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Readback data from ADF5902.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Store the read data.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int32_t adf5902_readback(struct adf5902_dev *dev, uint8_t reg_addr,
+			 uint32_t *data)
+{
+	int32_t ret;
+	uint32_t i;
+	uint8_t buff[ADF5902_BUFF_SIZE_BYTES];
+
+	*data = *data | reg_addr;
+
+	buff[0] = *data >> 24;
+	buff[1] = *data >> 16;
+	buff[2] = *data >> 8;
+	buff[3] = *data;
+
+	/* Write command */
+	ret = spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+	if (ret != SUCCESS)
+		return ret;
+
+	/* Send dummy value to get the data on DOUT */
+	buff[0] = ADF5902_SPI_DUMMY_DATA;
+	buff[1] = ADF5902_SPI_DUMMY_DATA;
+	buff[2] = ADF5902_SPI_DUMMY_DATA;
+	buff[3] = ADF5902_SPI_DUMMY_DATA;
+
+	ret = spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+	if (ret != SUCCESS)
+		return ret;
+
+	/* Concatenate received data Bytes */
+	for (i = 0; i < ADF5902_BUFF_SIZE_BYTES; i++)
+		*data = (*data << 8) | buff[i];
+
+	return ret;
+}
+
+/**
+ * @brief Compute ADF4350 RF VCO frequency parameters.
+ * @param dev - The device structure.
+ * @return Returns 0 in case of success or negative error code.
+ */
+static int32_t adf5902_vco_freq_param(struct adf5902_dev *dev)
+{
+
+	dev->ref_div_factor = 0;
+
+	do {
+		dev->ref_div_factor++;
+		dev->f_pfd = (dev->ref_in * (1 + dev->ref_doubler_en) /
+			      (dev->ref_div_factor *(1 + dev->ref_div2_en)));
+	} while (dev->f_pfd > ADF5902_MAX_FREQ_PFD);
+
+	dev->int_div = (uint16_t)(dev->rf_out / (dev->f_pfd * 2));
+
+	if (dev->int_div > ADF5902_MAX_INT_MSB_WORD)
+		return FAILURE;
+
+	dev->frac_msb = (uint16_t)(((dev->rf_out) * pow(2, 12) / (dev->f_pfd * 2)) -
```

Comment:
pow(2, x) is less efficient than shifting (1 << x), so I wouldn't use it in these cases.

Reply:
done.